c10/core/TensorImpl.h

```diff
@@ -907,6 +907,21 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   // because `shallow_copy_from()` is used for implementing functions such as `var.set_data(tensor)`, which
   // changes `var`'s tensor metadata and expects its `allow_tensor_metadata_change_` to be ignored.
 
+  /**
+   * One TensorImpl can be copied to another TensorImpl if they have the same
+   * type_id. The only two special cases (for legacy reason) are:
+   * CPUTensorId is compatible with CUDATensorId and SparseCPUTensorId is
+   * compatible with SparseCUDATensorId.
+   */
+  inline bool has_compatible_shallow_copy_type(const TensorTypeId& from) {
+    const TensorTypeId& self = type_id();
+    return (self == from) ||
+        ((self == CPUTensorId() || self == CUDATensorId()) &&
```

Comment:
I'm not sure I fully understand how HIP fits in to this, but I _think_ a HIP tenosr will have `HIPTensorId` even though its device is cuda and we want equivalent hip and cuda support.  I think @bddppq would know for sure.

Reply:
Here should add `self == HIPTensorId()` check as well (and similar for the other checks below)